p2p/nat/nat.go

```diff
@@ -29,6 +29,49 @@ import (
 	natpmp "github.com/jackpal/go-nat-pmp"
 )
 
+type Spec struct {
+	spec     string
+	instance Interface
+}
+
+func NewFromSpec(spec string) (*Spec, error) {
+	instance, err := Parse(spec)
+	if err != nil {
+		return nil, err
+	}
+	return &Spec{spec, instance}, nil
+}
+
+func MustFromSpec(spec string) *Spec {
+	instance, err := Parse(spec)
+	if err != nil {
+		panic(err)
+	}
+	return &Spec{spec, instance}
+}
+
+func (s *Spec) Implementation() Interface {
```

Comment:
Yeah I don't like this wording either. I was about to call it `Instance`, but that just felt too much like java `getInstance` singleton methods